### ISSUE TITLE
[LINUX remote pid] Both ends monitoring

### ISSUE THREAD

**Issue by anestisb** [google/honggfuzz#66]
The scenario is that you have a target system daemon-server that you want to fuzz test via a client cmd tool. Normally you would expect that monitoring/tracing only the daemon/server side would be enough. Unfortunately that is not the case for quite a few components in Android.

A common case is that the metadata (and a light verification) of the input data are processed locally from the client process (aka the pid group spawned from the fuzzer) and the heavy processing is performed from the daemon/server process (aka the remote long-lived pid group fuzzer attaches to). Since the client-side part will be incorporated also in privileged processes (which is not easy to fuzz at the first time) at a production setup, bugs at both sides are interesting for these scenarios.

Currently honggfuzz is configured to monitor either the spawned fuzzer-side or the remote pid. Can't do both. It would be really useful if such a feature can be implemented as part of the Linux arch_reapChild functionality.

I've made an attempt (https://github.com/anestisb/honggfuzz/commit/e7451149a1924f3d759aae2e7310eb5d3fcc92b3) a few weeks ago which seemed to work fine. Didn't bother to PR though since it was kind of ugly with lot's of duplicate code.

What are your thoughts for a more elegant implementation? Maybe this e2e pid monitoring can be splitted into two levels:
1. The attach/ptrace level: If chosen by the user ptrace attach tid groups on both sides
2. The feedback level: Which end will provide feedback data for the dynamic data prepare logic (both, fuzzer or remote pid)

The first is quite easy, although the second part might prove a bit tricky since it will require to offload the pid decisions from sancov & perf components and explicitly define it as part of the calling args.

**Comment by anestisb** [comment]
After playing with a few approaches it is not worth paying the cost of monitoring both ends from the same fuzzing campaign. Plus the fact that things get super complex/expensive when fuzzing with coverage/feedback since runtime data structures need to maintained twice.

For now I've solved this problem by running two campaigns in parallel: one attaching to long-lived server process and one spawning new fuzzer clients. I've written a small intermediate program to do the negotiation work between the two ends letting the long-lived fuzzer know when a new seed has been selected from the other end (pretty much an external preparation process wrapper).

Long story short, I don't think it's worth it to upstream such a feature. Thus closing.
